Improved Onboarding

digi0ps opened the pull request:
Adding new onboarding workflows for new users. [[Notion Doc]](https://www.notion.so/Improve-Onboarding-4a902c25e00543258684f6eab45c2668)

TODOs (Till Workflow 2):
- [x] Display new container with welcome message and checklist when user doesn't have any results at first.
- [x] Display checklist in alternative view when user has results.
- [x] Redirect to Wiki page of Memex on click of Checklist item 1.
- [x] Show first notification to make user highlight text.
- [x] Show second notification once user highlights text.
- [x] Keep popping up notification until user makes an annotation.
- [x] Open sidebar with "congrats" message once user makes an annotation.
- [x] Style Select Option Notification.
- [x] Style Congrats Message.
- [x] Tick off checklist in overview once the user is done with the steps.

TODOs (Workflow 3: Power Search):
- [x] Setup keys and redirect for demo 2
- [x] Display landing notification
- [x] Display go back to dashboard notification
- [x] Setup new overview tooltips
- [x] Display different tooltips based on stage
- [x] Style notification and tooltips

Problems/Bugs:
- [x] Previous notification doesn't close when new notification is created. ( @ShishKabab you have written code to avoid this behavior. But the [condition](https://github.com/WorldBrain/Memex/blob/9fad5d5bc5af70b32edbdd74f4c796889d87fe58/src/toolbar-notification/content_script/index.ts#L20) doesn't pass, even though am using the same `ToolbarNotification` instance.
- [x] Second Notification should get closed on a click which is made outside. Should I consider using `OnClickOutside` on ToolbarNotification?

oliversauter commented:
Just gave it a go :) 

So much smoother to onboard with this. Great! 
Apart from the stylings, which you already mentioned, all works well. 

Things that still need to be done: 
- [ ] once a user has seen the confirmation message that he has been finishing the onboarding of the annotations (and search respectively), the checkbox should be market as complete. 
- [ ] if you make the highlight (step 2)  in the bottom part of the page it does not show the tooltip, only unless it is inside the first part of the screen. http://recordit.co/3Wb2ek9lKT

digi0ps commented:
@oliversauter Alright, there's already will get it done!

oliversauter commented:
Some more feedback: 

- [ ] remove the first screen <img width="1145" alt="screen shot 2018-12-11 at 11 31 23" src="https://user-images.githubusercontent.com/7870039/49794658-62a86880-fd38-11e8-8a53-2a4c1c6a6157.png">
- [ ] don't try to load the results when you first open the app.

oliversauter commented:
Ah another thing I completely forgot: 
We need analytics events for each of the steps.

digi0ps commented:
>> don't try to load the results when you first open the app.
Users are directly taken to the Overview after install. So if we delete the initial onboarding screen, then they will be seeing the new big onboarding box first right?

oliversauter commented:
> Users are directly taken to the Overview after install. So if we delete the initial onboarding screen, then they will be seeing the new big onboarding box first right?

Yup. Also that correlates with the loading spinner that should not start > it does right now if you know what I mean. Basically it should only show the box.

digi0ps commented:
Got the logic and styling done for the Power Search stage, you can test them out now.

Few things that are remaining are (in the order of priority)
- [ ] Adding a shadow to the tooltip tip.
- [ ] Adding analytics
- [ ] Making onboarding tooltip position based on an element, instead of manually positioning it as it is being done right now.
- [ ] Converting `toolbar-notifications` components to tsx to remove the slew of errors.

There are some queries though, from a developer point of view @oliversauter.
1. In the [onboarding notion doc](https://www.notion.so/Improve-Onboarding-4a902c25e00543258684f6eab45c2668#600620042ed64495be64d4c64a49a97b) you have said: 
> Closes only on a little "X" that will be right next to the 'remember' (not there yet), or when user starts typing into the search box

> When user executed first search jump to 3.4

But as soon as the user starts typing, the query is executed and the results are shown. So I change to 3.4 tooltip when the user starts typing in the search box. Can you check it out and tell if it's this behaviour is fine?

2. Some of the previous onboarding components aren't being used now (afaik), what to do about them? Should I clean them up and have only the current onboarding components? ( Huge shoutout to @mukeshkharita for writing all the whole onboarding code, which helped me write the new onboarding components easily ).

3. Similarly what about the old tooltips? Temporarily I have just renamed the module to old-tooltip. But both nearly share the same Redux logic. If it is going to be used in the near future, I can put only the previous components into a separate folder and remove the others.

oliversauter commented:
Bugs: 
- [ ] If I have many tabs open I run into the issue that on new-install it takes a while for all tabs to be registered. Means if the Wikipedia tab wasn't registered yet it does not show the tooltips yet. @mohitsakhuja may already have a solution to that. In any case we should implement that the current active tab skips the queue for registering new tabs so this and other tab/shadowdom related activities are smooth. What I see though is that the ribbon is loaded immediately. Just not the tooltips. 
- [ ] Do not check of the "power search" to be done without the user having done their first actual search, including skipping through all the tooltips. Right now it is checked when I click on "get me back to dashboard" 
- [ ] add a "close" and "go back" button to each notification. No styling needed yet, just functionalities. "go back" should show the previous notification
- [ ] right now the "more filters" tooltip only appears if you also add a time filter to the "before" field. but needing to do that is not instructed. I'd say

oliversauter commented:
> But as soon as the user starts typing, the query is executed and the results are shown. So I change to 3.4 tooltip when the user starts typing in the search box. Can you check it out and tell if it's this behaviour is fine?

Yes. 

> Similarly what about the old tooltips?

Are those the ones that we had in the corner of the dashboard showing random tips?

> Some of the previous onboarding components aren't being used now (afaik), what to do about them? Should I clean them up and have only the current onboarding components? 

I think if they are not used, lets remove them. We can always go back in time if we need them again.

digi0ps commented:
> Are those the ones that we had in the corner of the dashboard showing random tips?

Yes.

> but it should only appear on the memex wikipedia page as part of the onboarding session.

Okay I will add a check for that.

digi0ps commented:
**Updates**
1. I have added analytics for everything except the overview tooltips in this [commit](https://github.com/WorldBrain/Memex/pull/644/commits/f2d37feb58a31e88d8b5fb2a1ac0eb17d9f9a6cc). Should I add analytics for them too?
2. I have also added the close and previous button for the tooltips. If the user clicks the close button, it closes that tooltip alone, the next tooltip can be triggered by the user. Is this the way it should be? or should it permanently close tooltip such that the next one doesn't appear?
3. Here is a [recorded screencast](http://recordit.co/m1UQjjROMd) of the behavior of the Time Filter tooltip.

digi0ps commented:
While refactoring the onboarding module, I noticed that the `FEATURES_INFO` from `overview/onboarding/constants` is being used to render the _FeatureInfo_ list in `options/tutorials`. I assume the tutorials component was used along with the previous onboarding component.

 Right now with the new onboarding, will this tutorial component be used? Or should I remove it? @oliversauter

oliversauter commented:
We will need to do a bit of a redesign of the component itself and add the guided tutorials there too, but the list is going to stay so that people can get more info easily.

digi0ps commented:
Alright. I will keep refactor the tutorial stuff out of onboarding and into it's own folder.

oliversauter commented:
> I have added analytics for everything except the overview tooltips in this commit. Should I add analytics for them too?

Yeah we would need to have analytics for each step in those tooltips too so we know when they break off. 

> I have also added the close and previous button for the tooltips. If the user clicks the close button, it closes that tooltip alone, the next tooltip can be triggered by the user. Is this the way it should be? or should it permanently close tooltip such that the next one doesn't appear?

It should actually let the next one appear. But maybe we call it  "next" then. Maybe we solve this better by adding 2 buttons < > to go back and forth, an a "finish" button to completely stop the tooltips (which then also should not check off the power search tutorial onboarding)

> Here is a recorded screencast of the behavior of the Time Filter tooltip.

oliversauter commented:
@digi0ps I got some workflow feedback from a UX person: 
- [ ] if the user clicks on "make your first power search" when the tutorial box is on the side (implying that there is already a result in the list) then it should not open the Memex wikipedia page again. Instead it should immediately jump to the search bar tooltip.
- [ ] if you create a memex.link first and then an annotation, it goes into infinity loading. 
- [ ] Add a remove button to the checklist if it is on the right side > but only on the right side element
- [ ] Remove the "import bookmarks" item and create a workflow for adding tags/collections via the popup. > will deliver screenshots

digi0ps commented:
> How did we solve it the last time?

I can't remember it right now, I will look into this. My guess is it must be the loading svg, I can try using data64 image instead of the svg.

> if the user clicks on "make your first power search" when the tutorial box is on the side (implying that there is already a result in the list) then it should not open the Memex wikipedia page again. Instead it should immediately jump to the search bar tooltip.

Did this. Feels much easier/intuitive now.

>  Instead put a button behind every element onhover saying "start". When the step is finished cross it out.

I don't quite get this. Do you mean to display Start over the Checklist text when the user hovers over it and strike it off when it's done?

oliversauter commented:
Another couple of things: 
- Can you make the < > and X icons? Its hard for me to change that. Just take any for now and I'll change it. 
- when all tutorial steps have been done, show a last screen with congrats and the ability to close the checklist popup (also add the ability to close the checklist popup with a little x > use icon instead of text 'x')

oliversauter commented:
> I don't quite get this. Do you mean to display Start over the Checklist text when the user hovers over it and strike it off when it's done?

Instead of the Checkbox, show an Icon in the front (i will add those). Gonna be 1 and 2, 3 etc. 
When a user hovers over the title or icon it would then change the text to "let's go" > make sure that the new `p` is the same length even when the next text is shorter, otherwise I expect some glitching. 
When the step is finished, strike it through and grey it out. When users hover over the striked through part show text "restart tutorial step"> which effectively would throw them into the onboarding steps again.

oliversauter commented:
well done  with the 3rd notification. 

I think I have a new clue why the problem appears that the notification does not load when I open it too fast with many tabs. 
My clue: the sidebar ribbon is loading just fine, so maybe it is 2 different processes there?

poltak commented:
> If I have many tabs open I run into the issue that on new-install it takes a while for all tabs to be registered. Means if the Wikipedia tab wasn't registered yet it does not show the tooltips yet. [...] What I see though is that the ribbon is loaded immediately. Just not the tooltips.

I have been playing around with this, but cannot reproduce. These are the steps I am performing:

1. open 18 tabs (takes ~10s for me to init and register all of these)
2. install ext
3. click the first "Make your first web annotation" onboarding choice
4. wiki page opens in new tab, with tooltips showing and onboarding tutorial ready to follow

Note: steps 3 and 4 are done while existing tab registration is still processing.

Do those steps lead the issue to reproduce for you guys? If not, please write out some steps that I can follow.

One important thing to note about how existing tab registration works:
It *only* deals with any tabs that were open when the ext is installed/updated. Any tabs created after the ext is installed (like that wiki page that opens after interacting with the onboarding) get registered as soon as they are opened. Even if the existing tabs registration is still processing, any newly opened tabs should get registered immediately (registering new tabs and existing tabs are separate processes).

**Loving the new onboarding so far btw!**

oliversauter commented:
Thanks for all the great work @digi0ps :) 

Will close this in favor of #679 as I could not push to your branch (seems I have no rights to do that) Couldn't figure it out how to add to the other PR, so closing this one.
